src/main/java/net/neoforged/neoforge/transfer/transaction/SnapshotJournal.java

```diff
@@ -0,0 +1,171 @@
+        for (int i = snapshots.size(); i <= nestingDepth; i++) {
+            snapshots.add(null);
+        }
+
+        if (snapshots.get(nestingDepth) == null) {
+            T snapshot = createSnapshot();
+            Objects.requireNonNull(snapshot, "Snapshot may not be null!");
+            snapshots.set(nestingDepth, snapshot);
+            transaction.addCloseCallback(this);
+        }
+    }
+
+    @Override
+    public void onClose(TransactionContext transaction, Transaction.Result result) {
+        //region Migration phase removal
+        //Neo: for testing and will be removed after deprecation period is over for handler reworks.
+        // This is to provide a quick way to give some metrics during the migration phase
+        int currentDepth = transaction.nestingDepth();
+        int max = Math.max(DEEPEST_LAYER, currentDepth);
+        if (max != DEEPEST_LAYER) {
+            DEEPEST_LAYER = max;
+            DEEPEST_SNAPSHOT = this;
+        }
+        //endregion
+
+        // Get and remove the relevant snapshot.
+        T snapshot = snapshots.remove(currentDepth);
+
+        //If the transaction was aborted, revert to snapshot
+        if (result.wasAborted()) {
+            // If the transaction was aborted, we just revert to the state of the snapshot.
+            revertToSnapshot(snapshot);
+            releaseSnapshot(snapshot);
+        } else if (currentDepth <= 0) {
+            //The transaction is the root.
+            originalState = snapshot;
+            transaction.addRootCloseCallback(this);
+        } else if (snapshots.get(currentDepth - 1) == null) {
+            // No snapshot yet, so move the snapshot one nesting level up.
+            snapshots.set(currentDepth - 1, snapshot);
+            // This is the first snapshot at this level: we need to call addCloseCallback.
+            transaction.getOpenTransaction(currentDepth - 1).addCloseCallback(this);
+        } else {
+            // There is already an older snapshot at the nesting level above, just release the newer one.
+            releaseSnapshot(snapshot);
+        }
+    }
+
+    /**
+     * @return The deepest nested layer from any transaction over the lifetime of the runtime. This is intended to identify some possible changes needed after migration. Not used outside Neo.
+     */
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    @ApiStatus.Internal
+    public static int getDeepestLayer() {
+        return DEEPEST_LAYER;
+    }
+
+    /**
+     * @return The deepest nested layer from any transaction over the lifetime of the runtime. This is intended to identify some possible changes needed after migration. Not used outside Neo.
+     */
+    @Deprecated(since = ResourceHandlerDeprecationHandling.MC_1_21_6)
+    @ApiStatus.Internal
+    public static String getDeepestSnapshot() {
+        if (DEEPEST_SNAPSHOT == null) return "Nothing";
+        return DEEPEST_SNAPSHOT.getClass().toString();
+    }
```

Comment:
Just very weird overall. I'm not sure what information we expect to gain from this.